lib/hotel.rb

```diff
@@ -0,0 +1,258 @@
+    end
+    #---exception handling---
+
+    # keys for date_reservation_has and id_reservation_hash
+    date_key = reservation.start_date
+    reservation_id = reservation.reservation_id
+
+    # iterate through dates from start_date to end_date
+    while date_key < reservation.end_date
+      # add room number to date_reservation hash
+      if @date_reservation_hash[date_key] == nil # first entry
+        @date_reservation_hash[date_key] = [reservation.room_number]
+      #add a value to the list of reservations
+      else
+        @date_reservation_hash[date_key] << reservation.room_number
+      end
+      # update the date
+      date_key += 1
+    end
+
+    # update id_reservation_hash
+    @id_reservation_hash[reservation_id] = [reservation.start_date, reservation.end_date, reservation.room_number, reservation.total_cost]
+
+  end
+
+  # available rooms method - returns array of available rooms and blocked for given date range
+  def get_available_rooms(start_year, start_month, start_day, end_year, end_month, end_day)
+    #initialize available_rooms
+    available_rooms = @room_numbers.dup
+    blocked_rooms = []
+
+    # make start_date and end_date objects
+    start_date = Date.new(start_year, start_month, start_day)
+    end_date = Date.new(end_year, end_month, end_day)
+
+    #create keys for hash
+    date_key = start_date
+
+    # loops through dates
+    while date_key < end_date
+
+      # loop through room numbers
+      available_rooms.each do |room|
+        #remove reserved rooms from available rooms array
+        if @date_reservation_hash[date_key] == nil
+          available_rooms = available_rooms
+        elsif @date_reservation_hash[date_key].include?(room)
+          available_rooms.delete(room)
+        end
+
+        #put blocked rooms in an array
+        if @date_block_hash[date_key] == nil
+          blocked_rooms = blocked_rooms
+        elsif @date_block_hash[date_key].include?(room) == true && blocked_rooms.include?(room) == false
+          blocked_rooms << room
+        end
+      end
+
+      date_key += 1
+    end
+
+    return [available_rooms, blocked_rooms]
+  end
+
+  # updates blocks hash with info about block: [start_date, end_date, rooms, discount]
+  def create_block(start_year, start_month, start_day, end_year, end_month, end_day, block_rooms, cost, discount)
+    # dates
+    start_date = Date.new(start_year, start_month, start_day)
+    end_date = Date.new(end_year, end_month, end_day)
+
+    # create an id for the block which is (100000*num_days + (start_date +  first room number)
+    num_days = end_date - start_date
+    block_id = ((num_days*100000) + start_date.to_time.to_i / (60 * 60 * 24) + block_rooms[0]).to_i
+
+    # check what rooms are available
+    available_rooms = self.get_available_rooms(start_year, start_month, start_day, end_year, end_month, end_day)[0]
+
+    blocked_rooms = self.get_available_rooms(start_year, start_month, start_day, end_year, end_month, end_day)[1]
```

Comment:
You end up calling `get_available_rooms` twice here.  Could this be done once and then you can pull out the two sub-arrays?